# Hidden Markov Modelling


Hidden Markov Modeling (HMM) is a statistical model used to describe a system where there are observable outcomes (observations) that are influenced by an underlying, unobservable state (hidden state). In an HMM, the model assumes that the system is a Markov process, meaning that the probability of transitioning from one state to another depends only on the current state and not on the sequence of events that preceded it (memoryless property).

# Components of an HMM

Hidden States: These are the unobservable states of the system. The system is assumed to be in one of these hidden states at any given time. Hidden states form a Markov chain, meaning that the probability of transitioning from one hidden state to another depends only on the current hidden state.
Observations: Observable outcomes or measurements that are influenced by the hidden state. Each hidden state is associated with a probability distribution over the possible observations. However, the specific observation emitted at each time step is determined probabilistically according to this distribution.
Transition Probabilities: The probabilities of transitioning from one hidden state to another. These probabilities define the dynamics of the system and are typically represented by a transition probability matrix.
Emission Probabilities: The probabilities of observing each possible outcome given the current hidden state. These probabilities define how likely each observation is under each hidden state and are typically represented by an emission probability matrix.

# Applications of Hidden Markov Models

Speech Recognition: Modeling phonemes as hidden states and audio features as observations.
Natural Language Processing: Modeling parts of speech or syntax as hidden states and words or tokens as observations.
Bioinformatics: Modeling DNA sequences with regions of hidden functionality (e.g., genes, regulatory elements) as hidden states and nucleotides as observations.
Finance: Modeling market regimes (e.g., bull, bear) as hidden states and asset prices as observations.

# Working Principle

Initialization: Initialize the model with initial parameters, such as initial state probabilities, transition probabilities, and emission probabilities.
Forward Algorithm: Compute the probability of observing a sequence of observations given the model parameters, accounting for all possible hidden state sequences that could have generated the observations.
Backward Algorithm: Compute the probability of being in a particular hidden state at a given time given the observed sequence, also accounting for all possible hidden state sequences.
Expectation-Maximization (EM) Algorithm (Optional): Refine the model parameters by iteratively estimating them based on the observed data, using the forward-backward algorithms to compute the expected counts of transitions and emissions.
Decoding: Determine the most likely sequence of hidden states that generated the observed sequence, known as the Viterbi algorithm.

In [ ]:
# Install packages

pip install hmmlearn

In [2]:
# Import libraries and objects

from hmmlearn import hmm
import numpy as np
import pandas as pd

In [ ]:
# Load Functions

execfile('C:/Python_Data_Sets/Functions 10_07_2023.py')

In [9]:
# Load Data

GTD = pd.read_csv('C:/R Portfolio/Global_Terrorism_Prediction/globalterrorismdb_0522dist.csv', 
                     encoding = 'latin1',
                     low_memory = False)
GTD_1 = pd.read_csv('C:/R Portfolio/Global_Terrorism_Prediction/globalterrorismdb_2021Jan-June_1222dist.csv', 
                     encoding = 'latin1',
                     low_memory = False)
GTD_combined = pd.concat([GTD, GTD_1], ignore_index = True)
GTD_combined

eventid  iyear  imonth  iday  approxdate  extended  resolution  \
0       197000000001   1970       7     2         NaN         0         NaN   
1       197000000002   1970       0     0         NaN         0         NaN   
2       197001000001   1970       1     0         NaN         0         NaN   
3       197001000002   1970       1     0         NaN         0         NaN   
4       197001000003   1970       1     0         NaN         0         NaN   
...              ...    ...     ...   ...         ...       ...         ...   
214661  202106300023   2021       6    30         NaN         0         NaN   
214662  202106300029   2021       6    30  06/30/2021         0         NaN   
214663  202106300030   2021       6    30         NaN         1  08/07/2021   
214664  202106300038   2021       6    30         NaN         0         NaN   
214665  202107020037   2021       6    30         NaN         0         NaN   

        country         country_txt  region  ...  \
0            58  Dominican Republic       2  ...   
1           130              Mexico       1  ...   
2           160         Philippines       5  ...   
3            78              Greece       8  ...   
4           101               Japan       4  ...   
...         ...                 ...     ...  ...   
214661        4         Afghanistan       6  ...   
214662      138             Myanmar       5  ...   
214663      147             Nigeria      11  ...   
214664       45            Colombia       3  ...   
214665      138             Myanmar       5  ...   

                                                 addnotes  \
0                                                     NaN   
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...                                                   ...   
214661                                                NaN   
214662                                                NaN   
214663                                                NaN   
214664                                                NaN   
214665  There is doubt that this incident meets terror...   

                                                   scite1  \
0                                                     NaN   
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...                                                   ...   
214661  "Gunmen blow up power pylon in Parwan," Afghan...   
214662  "Spring Revolution Daily News for 16-30 June 2...   
214663  "Boko Haram Releases Abducted Catholic Priest ...   
214664  "Two dead and one wounded after clashes betwee...   
214665  "Spring Revolution Daily News for 2-7 July 202...   

                                                   scite2  \
0                                                     NaN   
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...                                                   ...   
214661  "Provinces hit by blackout after power pylon d...   
214662                                                NaN   
214663  "Kidnapped Maiduguri Catholic Priest regains f...   
214664                                                NaN   
214665                                                NaN   

                                                   scite3  \
0                                                     NaN   
1                                                     NaN   
2                                                     NaN

In [11]:
# Process Data

GTD_New = preprocess_data(GTD_combined)
GTD_New

Year  Month  Day             Country                       Region  \
0       1970      7    2  Dominican Republic  Central America & Caribbean   
1       1970      0    0              Mexico                North America   
2       1970      1    0         Philippines               Southeast Asia   
3       1970      1    0              Greece               Western Europe   
4       1970      1    0               Japan                    East Asia   
...      ...    ...  ...                 ...                          ...   
214661  2021      6   30         Afghanistan                   South Asia   
214662  2021      6   30             Myanmar               Southeast Asia   
214663  2021      6   30             Nigeria           Sub-Saharan Africa   
214664  2021      6   30            Colombia                South America   
214665  2021      6   30             Myanmar               Southeast Asia   

        Province           City   Longitude   Latitude                Attack  \
0       National  Santo Domingo  -69.951164  18.456792         Assassination   
1        Federal    Mexico city  -99.086624  19.371887   HostageKidnapAttack   
2         Tarlac        Unknown  120.599741  15.478598         Assassination   
3         Attica         Athens   23.762728  37.997490            BombAttack   
4        Fukouka        Fukouka  130.396361  33.580412  InfrastructureAttack   
...          ...            ...         ...        ...                   ...   
214661    Parwan    Jangal Bagh   69.196838  35.054772            BombAttack   
214662      Shan           Muse   97.897143  23.986739   HostageKidnapAttack   
214663     Borno        Unknown   13.014035  11.572869   HostageKidnapAttack   
214664     Cauca        Unknown  -76.333069   3.104189            BombAttack   
214665      Shan    Kaungmutone   97.895734  23.974583            BombAttack   

                  Target                                              Group  \
0                Private                                             MANO-D   
1                GovtDip                 23rd of September Communist League   
2       JournalistsMedia                                         OtherGroup   
3                GovtDip                                         OtherGroup   
4                GovtDip                                         OtherGroup   
...                  ...                                                ...   
214661         Utilities                                         OtherGroup   
214662            EduIns                                         OtherGroup   
214663         RelFigIns                                         Boko Haram   
214664     UnknownTarget  Revolutionary Armed Forces of Colombia (FARC) ...   
214665          Business                                         OtherGroup   

             Weapon  Dead  Lethal  
0       OtherWeapon   1.0       1  
1       OtherWeapon   0.0       0  
2       OtherWeapon   1.0       1  
3        Explosives   0.0       0  
4        Incendiary   0.0       0  
...             ...   ...     ...  
214661   Explosives   0.0       0  
214662  OtherWeapon   1.0       1  
214663     Firearms   0.0       0  
214664   Explosives   0.0       0  
214665   Explosives   0.0       0  

[214666 rows x 15 columns]

In [12]:
# South Asia Region Data

SA_data = GTD_New[GTD_New['Region'] == 'South Asia']
SA_data

Year  Month  Day      Country      Region            Province  \
585     1970     11    1     Pakistan  South Asia               Sindh   
1186    1972      2   22        India  South Asia               Delhi   
1862    1973      5    1  Afghanistan  South Asia               Kabul   
2216    1974      2    2     Pakistan  South Asia               Sindh   
2704    1974     12    9     Pakistan  South Asia                NWFP   
...      ...    ...  ...          ...         ...                 ...   
214636  2021      6   29        India  South Asia      Madhya Pradesh   
214648  2021      6   30  Afghanistan  South Asia             Paktika   
214649  2021      6   30  Afghanistan  South Asia               Herat   
214650  2021      6   30     Pakistan  South Asia  Khyber_Pakhtunkhwa   
214661  2021      6   30  Afghanistan  South Asia              Parwan   

                    City  Longitude   Latitude               Attack  \
585              Karachi  67.143311  24.891115        Assassination   
1186           New Delhi  77.153336  28.585836            Hijacking   
1862               Kabul  69.147011  34.516895              Unknown   
2216             Karachi  67.143311  24.891115           BombAttack   
2704            Peshawar  71.537430  34.006004           BombAttack   
...                  ...        ...        ...                  ...   
214636           Bamhani  80.368033  22.478282  HostageKidnapAttack   
214648           Sharana  68.763192  33.125658           BombAttack   
214649  Ghoryan district  61.162500  34.182222           BombAttack   
214650           Dwa Toi  69.554427  32.871622   ArmedAssaultAttack   
214661       Jangal Bagh  69.196838  35.054772           BombAttack   

                  Target              Group  \
585              GovtDip         OtherGroup   
1186    AirportsAircraft       Palestinians   
1862    AirportsAircraft     Black December   
2216            Maritime  Muslim Guerrillas   
2704             GovtDip         OtherGroup   
...                  ...                ...   
214636           Private            Maoists   
214648    Transportation            Taliban   
214649           Private            Taliban   
214650          Military         OtherGroup   
214661         Utilities         OtherGroup   

                                                   Weapon  Dead  Lethal  
585     Vehicle (not to include vehicle-borne explosiv...   4.0       1  
1186                                           Explosives   0.0       0  
1862                                          OtherWeapon   0.0       0  
2216                                             Firearms   0.0       0  
2704                                           Explosives   0.0       0  
...                                                   ...   ...     ...  
214636                                           Firearms   1.0       1  
214648                                         Explosives   0.0       0  
214649                                         Explosives   2.0       1  
214650                                           Firearms   2.0       1  
214661                                         Explosives   0.0       0  

[56578 rows x 15 columns]

In [13]:
# Group Infequent Categories

SA_data_1 = group_infrequent_categories(SA_data)
SA_data_1 

Year  Month  Day      Country      Region            Province  \
585     1970     11    1     Pakistan  South Asia               Sindh   
1186    1972      2   22        India  South Asia       OtherProvince   
1862    1973      5    1  Afghanistan  South Asia       OtherProvince   
2216    1974      2    2     Pakistan  South Asia               Sindh   
2704    1974     12    9     Pakistan  South Asia       OtherProvince   
...      ...    ...  ...          ...         ...                 ...   
214636  2021      6   29        India  South Asia       OtherProvince   
214648  2021      6   30  Afghanistan  South Asia       OtherProvince   
214649  2021      6   30  Afghanistan  South Asia       OtherProvince   
214650  2021      6   30     Pakistan  South Asia  Khyber_Pakhtunkhwa   
214661  2021      6   30  Afghanistan  South Asia       OtherProvince   

             City  Longitude   Latitude               Attack       Target  \
585     OtherCity  67.143311  24.891115        Assassination  OtherTarget   
1186    OtherCity  77.153336  28.585836          OtherAttack  OtherTarget   
1862    OtherCity  69.147011  34.516895              Unknown  OtherTarget   
2216    OtherCity  67.143311  24.891115           BombAttack  OtherTarget   
2704    OtherCity  71.537430  34.006004           BombAttack  OtherTarget   
...           ...        ...        ...                  ...          ...   
214636  OtherCity  80.368033  22.478282  HostageKidnapAttack      Private   
214648  OtherCity  68.763192  33.125658           BombAttack  OtherTarget   
214649  OtherCity  61.162500  34.182222           BombAttack      Private   
214650  OtherCity  69.554427  32.871622   ArmedAssaultAttack     Military   
214661  OtherCity  69.196838  35.054772           BombAttack  OtherTarget   

             Group       Weapon  Dead  Lethal  
585     OtherGroup  OtherWeapon   4.0       1  
1186    OtherGroup   Explosives   0.0       0  
1862    OtherGroup  OtherWeapon   0.0       0  
2216    OtherGroup     Firearms   0.0       0  
2704    OtherGroup   Explosives   0.0       0  
...            ...          ...   ...     ...  
214636  OtherGroup     Firearms   1.0       1  
214648     Taliban   Explosives   0.0       0  
214649     Taliban   Explosives   2.0       1  
214650  OtherGroup     Firearms   2.0       1  
214661  OtherGroup   Explosives   0.0       0  

[56578 rows x 15 columns]

# Reshaping the Data

SA_data_1['Lethal'].values: This extracts the values from the Lethal column in the DataFrame SA_data_1. The Lethal column presumably contains binary data indicating whether each recorded attack resulted in death (1) or not (0).
.reshape(-1, 1): This method reshapes the data into a two-dimensional array (a matrix). The -1 specifies that the number of rows should be inferred based on the length of the array, ensuring that all elements are included. The 1 indicates that there should be one column. This reshaping is necessary because hmmlearn requires input data to be in a two-dimensional array format even if there is only one feature per observation (as is the case here with the Lethal column).

In [14]:
# Reshaping the Data

X = SA_data_1['Lethal'].values.reshape(-1, 1)  # Reshape needed for hmmlearn

# Specifying Sequence Lengths

In [ ]:
lengths = [len(X)]  # All data is one big sequence

lengths = [len(X)]: This line creates a list containing a single element, which is the length of the array X. In the context of hmmlearn, the lengths array is used to specify the lengths of individual sequences when you have multiple sequences concatenated into the dataset X. In this case, it indicates that there is just one continuous sequence, encompassing the entire dataset.

In [15]:
# Define a Gaussian HMM
# Assume 2 hidden states; adjust based on domain knowledge or experimentation
model = hmm.GaussianHMM(n_components = 2, 
                        covariance_type = "diag",
                        n_iter = 1000)

# Fit model
model.fit(X)

GaussianHMM(n_components=2, n_iter=1000)

# Transition Probabilities

In [16]:
# Predict hidden states
hidden_states = model.predict(X)

# Show transition probabilities
print("Transition probabilities:")
print(model.transmat_)

# Show the means of each hidden state
print("Means of each hidden state:")
print(model.means_)

Transition probabilities:
[[0.63068736 0.36931264]
 [0.44167993 0.55832007]]
Means of each hidden state:
[[1.]
 [0.]]


Interpretation of the output:
From State 0 to State 0: There is a 63.07% chance that the system will remain in State 0 after being in State 0.
From State 0 to State 1: There is a 36.93% chance that the system will move to State 1 after being in State 0.
From State 1 to State 0: There is a 44.17% chance that the system will move to State 0 after being in State 1.
From State 1 to State 1: There is a 55.83% chance that the system will remain in State 1 after being in State 1.

# Purpose and Context

The purpose of these steps is to format the data correctly and inform the model about the structure of the dataset. By reshaping the data into a 2D array and specifying the length of the sequence, you are setting up the data in a way that hmmlearn can process effectively.

The reshaping is critical because machine learning models, including those in hmmlearn, often expect data to be in a format where each row represents a sample (or observation) and each column represents a feature. Even if there is only one feature (like here with Lethal), the data still needs to be explicitly presented as a matrix with one column for compatibility.
Specifying the sequence length is essential in scenarios where multiple sequences are being modeled together. Even though you have only one sequence in this instance, the library still requires this information to understand how the data is structured.
This setup essentially tells hmmlearn that it is dealing with a single sequence of attack data, with each data point representing whether an attack was lethal. This can be useful for analyzing temporal or order-dependent patterns in how attack outcomes occur if the data is sequential (i.e., ordered by time).

# Conclusion

Hidden Markov Models are powerful tools for modeling sequential data and are widely used in various fields, ranging from natural language processing to bioinformatics and finance. Their ability to capture dependencies between observations and to model complex systems with hidden structure makes them versatile and applicable to a wide range of problems.